File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque


def comp(a, b):
    cnt = 0
    
    for i in range(len(a)):
        if a[i] != b[i]:
            cnt += 1
    
    return cnt == 1


def bfs(begin, target, words):
    que = deque()
    que.append((begin, 0))
    
    visit = {}
    
    while que:
        cur, cnt = que.popleft()
        
        for word in words:
            if visit.get(word) is not None or not comp(word, cur):
                continue
                
            if word == target:
                return cnt + 1
            
            visit[word] = True
            que.append((word, cnt + 1))
            
    return 0

def solution(begin, target, words):
    return bfs(begin, target, words)

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    
    class Word{
        int dist;
        long visit;
        String val;
        
        public Word(int dist, long visit, String val){
            this.dist = dist;
            this.visit = visit;
            this.val = val;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        Queue<Word> que = new ArrayDeque<>();
        que.add(new Word(0, 0, begin));
        
        Map<String, Boolean> visit = new HashMap<>();
        visit.put(begin, true);
        
        while(!que.isEmpty()){
            Word cur = que.poll();
            
            for(String word : words){
                if(visit.containsKey(word) || !onlyOneDifferent(word, cur.val))
                    continue;
                
                if(target.equals(word))
                    return cur.dist + 1;
                
                que.add(new Word(cur.dist + 1, 0, word));
                visit.put(word, true);
            }
        }
        
        return 0;
    }
    
    private boolean onlyOneDifferent(String word1, String word2){
        int cnt = 0;
        
        for(int i=0 ; i<word1.length() ; i++){
            if(word1.charAt(i) != word2.charAt(i))
                cnt++;
            
            if(cnt >= 2)
                break;
        }
        
        return cnt == 1;
    }
}